peps/pep-0735.rst

```diff
@@ -0,0 +1,476 @@
+``packages``, which is a table. All other keys in ``requirements`` are reserved
+for future use.
+
+The ``packages`` table contains an arbitrary number of user-defined keys, each of
+which has, as its value, a list of requirements specifiers (defined below).
+These keys must match the following regular expression:
+``[a-z0-9][a-z0-9-]*[a-z0-9]``. Meaning that they must be all lower-case
+alphanumerics, with ``-`` allowed only in the middle, and at least
+two characters long. These requirements are chosen so that the normalization
+rules used for pypi package names do not apply, as the names are already
+normalized.
+
+Requirements specifiers will use a definition which extends PEP 508. Therefore,
+it is first necessary to define a string format, a "PEP 735 Dependency". These
+are defined as strings in one of the three following formats:
+
+* a PEP 508 specification, e.g., ``numpy>1``
+* a comma-delimited list of Dependency Group names, enclosed in square
+  brackets, e.g., ``[test,docs]`` to refer to other Dependency Groups named
+  ``test`` and ``docs``
+* a single dot, ``.``, which refers to the current project as a package
+  (similar to ``pip install .``)
+* a single dot, followed by square brackets enclosing a list of extra names,
+  which refers to the current project as a package including some extras, e.g.,
+  ``.[mysql]`` to refer to the current package with its ``mysql`` extra
+
+Requirement specifiers can now be defined as one of the following:
+
+* A string, which is a valid PEP 735 Dependency Specifier. e.g., ``numpy>1``
+* An object, which has exactly one key, ``spec``, which is a valid PEP 735
+  Dependency Spec. e.g., ``{spec = "numpy>1"}``
+
+Any additional keys in a requirement specifier object are reserved for future
+use.
+
+The following is an example of a ``pyproject.toml`` section which uses this to
+define four dependency groups: ``test``, ``docs``, ``typing``, and
+``typing-test``:
+
+.. code:: toml
+
+    [requirements.packages]
+    test = ["pytest", "coverage", "."]
+    docs = ["sphinx", "sphinx-rtd-theme"]
+    typing = ["mypy", "types-requests", ".[types]"]
+    typing-test = ["[typing,test]", "useful-types"]
+
+Note how ``test`` and ``typing`` are able to refer to the current package
+while ``docs`` does not. This reflects the ability of Dependency Groups to be used
+in the same manner as extras, adding to dependencies, or completely
+independently. ``typing-test`` is defined as a union of two existing groups,
+plus an additional package. ``typing`` includes an extra, ``types``.
+
+Implementation Requirements
+---------------------------
+
+Build backends MUST NOT include dependency group data in built distributions.
+
+Tools which support Dependency Groups MUST support both string and object
+representations of requirements.
+
+When unrecognized keys are encountered in requirement specifiers or the
+``requirements`` table, tools MUST NOT fail. They MAY emit warnings.
```

Comment:
```suggestion
``requirements`` table, tools MUST NOT fail. They SHOULD emit warnings.
```

I'm not quite sure how I feel about the "MUST NOT fail" part. Why should people be allowed to specify invalid group names?

Reply:
IMO, if we aren't comfortable with the "MUST NOT fail" semantics, we should drop the object specification from this version of the spec. To me, the two only make sense together -- if tools are able to fail on new keys being added, I'm not sure anything is gained by including the object format from the outset.

Reply:
Having reread the comment about "surprise success" below and thought about it a moment more, I'm not sure I'm as firm in this belief as I was only a few minutes ago. Even so, I think that if the "MUST NOT" gets removed, we/I should strongly consider removing the object specification.

Reply:
Since all of the language about handling unrecognized keys has been removed, I'm marking this resolved.